File: app/models/api/integrations.py
# ruff: noqa: F401
from __future__ import annotations

from enum import StrEnum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.constants import TWEET_SUGGESTION_MODEL
from app.models.api.base import UTCDateTime
from app.models.api.pagination import PaginationMetadata
from app.models.contracts import (
    ContentClassification,
    ContentStatus,
    ContentType,
    DeleteStatus,
    IntegrationDisconnectStatus,
    SummaryKind,
    SummaryVersion,
    UserLlmProvider,
)


class XOAuthStartRequest(BaseModel):
    """Request to begin X OAuth flow."""

    twitter_username: str | None = Field(default=None, max_length=50)


class XOAuthStartResponse(BaseModel):
    """Response payload for X OAuth start."""

    authorize_url: str
    state: str
    scopes: list[str] = Field(default_factory=list)


class XOAuthExchangeRequest(BaseModel):
    """Request to exchange an X OAuth authorization code."""

    code: str = Field(..., min_length=1, max_length=4096)
    state: str = Field(..., min_length=1, max_length=255)


class XConnectionResponse(BaseModel):
    """Current X integration connection state for a user."""

    provider: str
    connected: bool
    is_active: bool
    provider_user_id: str | None = None
    provider_username: str | None = None
    scopes: list[str] = Field(default_factory=list)
    last_synced_at: UTCDateTime | None = None
    last_status: str | None = None
    last_error: str | None = None
    twitter_username: str | None = None


class IntegrationDisconnectResponse(BaseModel):
    """Response for integration disconnect actions."""

    status: IntegrationDisconnectStatus = IntegrationDisconnectStatus.DISCONNECTED
    provider: str = "x"


class UserLlmIntegrationResponse(BaseModel):
    """User-managed LLM integration summary."""

    provider: UserLlmProvider
    configured: bool
    updated_at: UTCDateTime | None = None


class UpsertUserLlmIntegrationRequest(BaseModel):
    """Request to store a user-managed LLM provider key."""

    api_key: str = Field(..., min_length=1, max_length=4096)


class UserLlmIntegrationTestResponse(BaseModel):
    """Response for validating presence of a user-managed LLM key."""

    provider: UserLlmProvider
    ok: bool


class DeleteUserLlmIntegrationResponse(BaseModel):
    """Response for deleting a user-managed LLM provider key."""

    status: DeleteStatus = DeleteStatus.DELETED
    provider: str

File: rust/crates/newsly-contracts/src/integrations.rs
use chrono::{DateTime, Utc};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, JsonSchema, ToSchema)]
#[serde(deny_unknown_fields)]
pub struct XOAuthStartRequest {
    pub twitter_username: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
pub struct XOAuthStartResponse {
    pub authorize_url: String,
    pub state: String,
    #[serde(default)]
    pub scopes: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, JsonSchema, ToSchema)]
#[serde(deny_unknown_fields)]
pub struct XOAuthExchangeRequest {
    pub code: String,
    pub state: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema, ToSchema)]
pub struct XConnectionResponse {
    pub provider: String,
    pub connected: bool,
    pub is_active: bool,
    pub provider_user_id: Option<String>,
    pub provider_username: Option<String>,
    #[serde(default)]
    pub scopes: Vec<String>,
    #[schemars(with = "Option<String>")]
    #[schema(value_type = Option<String>, format = DateTime)]
    pub last_synced_at: Option<DateTime<Utc>>,
    pub last_status: Option<String>,
    pub last_error: Option<String>,
    pub twitter_username: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum IntegrationDisconnectStatus {
    Disconnected,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
pub struct IntegrationDisconnectResponse {
    pub status: IntegrationDisconnectStatus,
    pub provider: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum UserLlmProvider {
    Anthropic,
    Openai,
}

impl UserLlmProvider {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Anthropic => "anthropic",
            Self::Openai => "openai",
        }
    }
}

impl TryFrom<&str> for UserLlmProvider {
    type Error = String;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "anthropic" => Ok(Self::Anthropic),
            "openai" => Ok(Self::Openai),
            other => Err(format!("unsupported user LLM provider {other:?}")),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema, ToSchema)]
pub struct UserLlmIntegrationResponse {
    pub provider: UserLlmProvider,
    pub configured: bool,
    #[schemars(with = "Option<String>")]
    #[schema(value_type = Option<String>, format = DateTime)]
    pub updated_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, JsonSchema, ToSchema)]
#[serde(deny_unknown_fields)]
pub struct UpsertUserLlmIntegrationRequest {
    pub api_key: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
pub struct UserLlmIntegrationTestResponse {
    pub provider: UserLlmProvider,
    pub ok: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum DeleteStatus {
    Deleted,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
pub struct DeleteUserLlmIntegrationResponse {
    pub status: DeleteStatus,
    pub provider: String,
}
